src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -8,16 +8,117 @@ namespace Microsoft.Azure.ServiceBus.Core
+
+        readonly ConcurrentExpiringSet<Guid> requestResponseLockedMessages;
+        readonly string entityPath;
+        readonly bool isSessionReceiver;
         readonly TimeSpan operationTimeout;
         readonly object messageReceivePumpSyncLock;
+        readonly bool ownsConnection;
+
         int prefetchCount;
         long lastPeekedSequenceNumber;
         MessageReceivePump receivePump;
         CancellationTokenSource receivePumpCancellationTokenSource;
 
+        public MessageReceiver(
+            ServiceBusConnectionStringBuilder connectionStringBuilder,
+            ReceiveMode receiveMode = ReceiveMode.PeekLock,
+            RetryPolicy retryPolicy = null)
+            : this(connectionStringBuilder.GetNamespaceConnectionString(), connectionStringBuilder.EntityPath, receiveMode, retryPolicy)
+        {
+        }
+
+        public MessageReceiver(
+            string connectionString,
+            string entityPath,
+            ReceiveMode receiveMode = ReceiveMode.PeekLock,
+            RetryPolicy retryPolicy = null)
+            : this(entityPath, new ServiceBusNamespaceConnection(connectionString), receiveMode, retryPolicy)
+        {
+            if (string.IsNullOrWhiteSpace(connectionString))
+            {
+                throw Fx.Exception.ArgumentNullOrWhiteSpace(connectionString);
+            }
+            if (string.IsNullOrWhiteSpace(entityPath))
+            {
+                throw Fx.Exception.ArgumentNullOrWhiteSpace(entityPath);
+            }
+
+            this.ownsConnection = true;
+        }
+
+        public MessageReceiver(
+            string entityPath,
+            ServiceBusConnection serviceBusConnection,
+            ReceiveMode receiveMode = ReceiveMode.PeekLock,
+            RetryPolicy retryPolicy = null)
+            : base(nameof(MessageSender) + StringUtility.GetRandomString(), retryPolicy ?? RetryPolicy.Default)
+        {
+            this.ReceiveMode = receiveMode;
+            this.operationTimeout = serviceBusConnection.OperationTimeout;
+            this.messageReceivePumpSyncLock = new object();
+            this.entityPath = entityPath;
+            this.ServiceBusConnection = serviceBusConnection;
+            this.PrefetchCount = prefetchCount;
+            var tokenProvider = TokenProvider.CreateSharedAccessSignatureTokenProvider(serviceBusConnection.SasKeyName, serviceBusConnection.SasKey);
+            this.CbsTokenProvider = new TokenProviderAdapter(tokenProvider, serviceBusConnection.OperationTimeout);
+            this.ReceiveLinkManager = new FaultTolerantAmqpObject<ReceivingAmqpLink>(this.CreateLinkAsync, this.CloseSession);
+            this.RequestResponseLinkManager = new FaultTolerantAmqpObject<RequestResponseAmqpLink>(this.CreateRequestResponseLinkAsync, this.CloseRequestResponseSession);
+            this.requestResponseLockedMessages = new ConcurrentExpiringSet<Guid>();
+        }
+
+        internal MessageReceiver(
```

Comment:
>internal MessageReceiver( [](start = 8, length = 25)

there seems to be a disconnect  between the public and the internal constructors. 
You can have multiple constructors, but all of them should essentially reach to the same constructor which initializes all properties of the class.